### DISCUSSION TITLE
Keyboard Shortcuts

### DISCUSSION THREAD

**Discussion by Lurkars** [Lurkars/gloomhavensecretariat#90] [Ideas]
Never used them on old helper, but got request for zooming with arrow keys on BGG and now thought, we can collect ideas for good shortcuts!

[Currently implemented](https://github.com/Lurkars/gloomhavensecretariat/blob/main/resources/keyboard-shortcuts.md)

**Comment by jgwehr** [comment]
I like the idea of `space` as a "next major action", moving things forward when conditions are met. Looks like you have `tab` and `shift tab` doing something similar (very impressive)! For example, starting from a pre-draw state:

1. Perform **Draw** if all characters have an initiative (+ any other conditions - any settings)
2. for each valid figure: apply conditions, etc. then **Draw Attack Modifier** if conditions are met (one or more standee numbers; not dead; settings enabled; etc)
3. **Draw Attack Modifier** when needed, such as rolling or Advantage/Disadvantage
4. **Close Drawn Attack Modifier(s)** so that the user can modify health, loot, etc
5. **End Turn** and **next**
6. finally, **Next Round**

Alternatively, `space`: <Draw Monster Ability>

-------
Other ideas: 

Not so sure about these since they might conflict with Initiative
- `1`: Cycle **Fire Element**
- `2`: Cycle **Ice Element**
- `3`: Cycle **Wind Element**
- `4`: Cycle **Earth Element**
- `5`: Cycle **Light Element**
- `6`: Cycle **Dark Element**

- `+` and `-` as alternates for **Zoom** (closer to most browsers' ctrl++ and cltr+-)
- `Arrow Up` and `Arrow Down` for **Health**, affecting the current figure
- `Arrow Left` and `Arrow Right` for **XP**? affecting the current figure
- alternate the arrow keys: `[` and `]` for HP, `;` and `'` for XP, `,` and `.` for coins

- `s` `i`  `d` `w` `m` `p` `a` for conditions? these seem tricky
- `x` for **kill** affecting the current figure
- `b` and `c` for blessing and curse

**Reply by Lurkars** [reply]
First, sorry I didn't update this, now added the current implemented shortcuts.

For your suggestions: some of them sounds a bit complicated. For others, you always need a "selected figure", like adding damage or set condition, so most of the time it won't be the active figure and app can't know which selected by now. So two ideas here: first add some shortcuts only when 

menu is open like: 
- 0-9 keys for toggling condition
- c and b for bless curse
- d or k for kill
- disable zoom shortcut and change health written arrow key

The other solution would be more complicated, to have a selected state on one or more figures. But I think that won't come quickly.

**Reply by jgwehr** [reply]
Totally understood! I like the updates in your original post. Zoom etc was working well last night when we played. In case you don't see it on reddit: the new 0.60 worked great on my docker containers, I'll post if I see any issues in the future.

**Reply by tomas-pals** [reply]
Wow I did not realise all of these shortcuts were there, I've been using the mouse way too much.
I would love to be able to modify the HP and conditions of any figure easily though since that's what you spend a lot of your time doing. I understand the issue with _selecting a figure_ which isn't the active figure. I had an idea though which is kind of inspired by the Jetbrains plugin [_AceJump_](https://plugins.jetbrains.com/plugin/7086-acejump). 

The idea would then be that you could press `O` (for _Open menu_) to go into figure selection mode. The UI would then show a number next to every figure, and by pressing that number you would then open that figures menu allowing for changes to all of its state
![CleanShot 2023-09-25 at 22 02 42@2x](https://github.com/Lurkars/gloomhavensecretariat/assets/2796248/28b2a0de-75c3-4c59-8c32-01b39ccc39c7)
So pressing `4` in this case would open the first _Hound_
<img width="733" alt="CleanShot 2023-09-25 at 22 07 17@2x" src="https://github.com/Lurkars/gloomhavensecretariat/assets/2796248/7c2adaac-b847-430f-adeb-6030c125590b">

**Reply by Lurkars** [reply]
Yes, good idea. So I will add shortcut for enabling numeric select and then one can select with numeric key input. If count is > 10, there is a 1s timeout to press second key like "1" , "5" for selecting 15. If entering any invalid number, selecting is cancelled.

**Comment by Lurkars** [comment]
So only change I made now is use key `s` for select instead of `o` for open. You can test the new shortcut in [v0.77.8](https://github.com/Lurkars/gloomhavensecretariat/releases/tag/v0.77.8). Happy about feedback.

**Reply by tomas-pals** [reply]
Wow well done! I can't believe how quickly you got this out 😁 It was exactly what I was thinking of.
I really want to start contributing as well, have done Angular for work previously so should be easy to pick up. Keep up the good work!
